### PR TITLE
Write guide on how to evolve the strategic merge patch format.

### PR DESCRIPTION
**Pull request by pwittrock**

### CODE REVIEW

File: contributors/devel/strategic-merge-patch.md [hunk 1]
```diff
@@ -3,7 +3,9 @@ Strategic Merge Patch
 
 # Background
 
-TODO: @pwittrock complete this section
+Kubernetes supports a customized version of JSON merge patch called strategic merge patch.  This
+patch format is used by `kubectl apply` and `kubectl edit`, and contains
```
Comment (mengqiy) [inline contributors/devel/strategic-merge-patch.md]: Should we mention `kubectl patch` here?

File: contributors/devel/strategic-merge-patch.md [hunk 2]
```diff
@@ -209,15 +211,115 @@ In an erroneous case, the set may be created with duplicates. Deleting an
...
+for the old format.
+
+### Client version skew
+
+Because the server does not publish which patch versions it supports,
+and it silently ignores patch directives that it does not recognize,
+new patches should behave correctly when sent to old servers that
+may not support all of the patch directives.
+
+`kubectl` is a client that must support talking to older and newer server versions,
```
Comment (mengqiy) [inline contributors/devel/strategic-merge-patch.md]: I think we should also say that patch from other old clients should work correctly on the new server. This guarantees we are not breaking APIs.

File: contributors/devel/strategic-merge-patch.md [hunk 3]
```diff
@@ -209,15 +211,115 @@ In an erroneous case, the set may be created with duplicates. Deleting an
...
+This example won't work, because old patch formats will contain data that is now
+considered required.  To support this, introduce a new directive to guard the
+new patch format.
+
+#### 2. Add support for new directives in the patch format
+
+- Optional directives may be introduced to change how the patch is applied by the server - **backwards compatible** (old patch against newer server).
+  - *Old servers will ignore directives they do not recognize*
+  - May control how the patch is applied
+  - May contain optional metadata such as a list of elements to delete from a list that was not present before
```
Comment (mengqiy) [inline contributors/devel/strategic-merge-patch.md]: >that was not present before

I feel this clause makes this sentence not very clear. May we should rewording it.

File: contributors/devel/strategic-merge-patch.md [hunk 4]
```diff
@@ -209,15 +211,115 @@ In an erroneous case, the set may be created with duplicates. Deleting an
...
+  - Must NOT impose new requirements on the old patch format
+
+- New patch requests should be a superset of old patch requests - **forwards compatibile** (newer patch against older server)
```
Comment (mengqiy) [inline contributors/devel/strategic-merge-patch.md]: Typo: compatible

File: contributors/devel/strategic-merge-patch.md [hunk 5]
```diff
@@ -209,15 +211,115 @@ In an erroneous case, the set may be created with duplicates. Deleting an
...
+  - unmodified fields for Kind foo are required in patch request
+
+This example won't work, because old patch formats will contain data that is now
+considered required.  To support this, introduce a new directive to guard the
+new patch format.
+
+#### 2. Add support for new directives in the patch format
+
+- Optional directives may be introduced to change how the patch is applied by the server - **backwards compatible** (old patch against newer server).
+  - *Old servers will ignore directives they do not recognize*
```
Comment (mengqiy) [inline contributors/devel/strategic-merge-patch.md]: > **backwards compatible** (old patch against newer server)

I guess we don't need this line here. This section is about new server. And this line has been  used below.

File: contributors/devel/strategic-merge-patch.md (hunk 1 above)
Comment (pwittrock) [inline contributors/devel/strategic-merge-patch.md]: Done

File: contributors/devel/strategic-merge-patch.md (hunk 2 above)
Comment (pwittrock) [inline contributors/devel/strategic-merge-patch.md]: Added this to the previous section

File: contributors/devel/strategic-merge-patch.md (hunk 5 above)
Comment (pwittrock) [inline contributors/devel/strategic-merge-patch.md]: Good point

File: contributors/devel/strategic-merge-patch.md (hunk 3 above)
Comment (pwittrock) [inline contributors/devel/strategic-merge-patch.md]: Reworded

File: contributors/devel/strategic-merge-patch.md (hunk 4 above)
Comment (pwittrock) [inline contributors/devel/strategic-merge-patch.md]: Done

**Comment by pwittrock** [comment]
PTAL

**Comment by adohe** [comment]
Overall looks good to me. @kubernetes/kubectl-maintainers anyone would like to take a look?

**Comment by mengqiy** [comment]
Can we merge this?

**Comment by pwittrock** [comment]
@mengqiy Thanks for reminding me

**Pull request merged**